### DISCUSSION TITLE
How do I stop hex colors shorter than 6 characters from being highlighted

### DISCUSSION THREAD

**Discussion by akinsho** [uga-rosa/ccc.nvim#77] [Q&A] [answered]
Having searched through similar issues, most closely related of which is https://github.com/uga-rosa/ccc.nvim/issues/14. I couldn't/cannot get this plugin to stop highlighting any numbers preceded by a # e.g. `#236` for example which could be a PR or issue number on GH and not a hex color. In general I don't want it highlight any hex colors that aren't >= 6.

I tried using the `exclude_pattern` option mentioned in that issue (it's undocumented btw, just mentioning so you're aware) for example:
```lua
        hex = {
          '%s#%x%x%x%s+',
          '[%w_]#%x%x%x',
          '[%w_]#%x%x%x%x%x%x',
        },
```
but the following text in a `lua` file with the highlighter enable still highlights ` -- partial fix for #391` the `#391` is still highlighted unfortunately.

I'm pretty sure that lua pattern works as I've tested that it would match the pattern but it doesn't stop the highlighting.

**Comment by uga-rosa** [comment, marked as answer]
`exclude_pattern` option is now obsolete. It was intended to specify word boundaries, and I have decided that regular expressions `\<`, `\>` and `[[:keyword:]]` are sufficient for this purpose.
Everything passed to pickers in `ccc.setup()` is an object, and manipulating this entity will accomplish the goal. Specifically, HexPicker has a field called pattern, which can be rewritten.
```lua
local ccc = require("ccc")

local HexPicker = ccc.picker.hex
HexPicker.pattern = {
  [=[\v%(^|[^[:keyword:]])\zs#(\x\x)(\x\x)(\x\x)>]=],
  [=[\v%(^|[^[:keyword:]])\zs#(\x)(\x)(\x)>]=],
}

ccc.setup({
  -- Other options
  -- ...
  pickers = {
    HexPicker,
    ccc.picker.css_rgb,
    ccc.picker.css_hsl,
    ccc.picker.css_hwb,
    ccc.picker.css_lab,
    ccc.picker.css_lch,
    ccc.picker.css_oklab,
    ccc.picker.css_oklch,
  },
})
```

**Reply by akinsho** [reply]
@uga-rosa thanks that's very helpful. Can this be set on a per filetype basis out of curiosity? this isn't something I need right now but if working in a non-lua file for example which for me is exclusively in the neovim so terminal context then I don't want anything other than 6 digit hex codes but if working on something for the browser e.g. js/ts then I might want to have 8 digit plus hex codes?

Just a question for curiosity not an immediate requirement

**Reply by uga-rosa** [reply]
I've mentioned it in another thread, implementation of picker/hex is [here](https://github.com/uga-rosa/ccc.nvim/blob/main/lua/ccc/picker/hex.lua). If you are using LuaLS, please also refer to the annotations.

How about branching in parse_color?
```lua
local ccc = require("ccc")

local HexPicker = ccc.picker.hex

local MyHexPicker = {
  pattern = {
    default = {
      [=[\v%(^|[^[:keyword:]])\zs#(\x\x)(\x\x)(\x\x)>]=],
      [=[\v%(^|[^[:keyword:]])\zs#(\x)(\x)(\x)>]=],
    },
    javascript = {
      [=[\v%(^|[^[:keyword:]])\zs#(\x\x)(\x\x)(\x\x)>]=],
      [=[\v%(^|[^[:keyword:]])\zs#(\x)(\x)(\x)>]=],
      [=[\v%(^|[^[:keyword:]])\zs#(\x\x)(\x\x)(\x\x)(\x\x)>]=],
      [=[\v%(^|[^[:keyword:]])\zs#(\x)(\x)(\x)(\x)>]=],
    },
  },
}

function MyHexPicker:parse_color(s, init)
  local ft = vim.bo.filetype
  local pattern = self.pattern.default
  if vim.tbl_contains({ "javascript", "typescript" }, ft) then
    pattern = self.pattern.javascript
  end
  HexPicker.pattern = pattern
  return HexPicker:parse_color(s, init)
end

ccc.setup({
  pickers = {
    MyHexPicker,
    ccc.picker.css_rgb,
    ccc.picker.css_hsl,
    ccc.picker.css_hwb,
    ccc.picker.css_lab,
    ccc.picker.css_lch,
    ccc.picker.css_oklab,
    ccc.picker.css_oklch,
  },
})
```

**Reply by akinsho** [reply]
Thanks @uga-rosa good to have the reference.

**Comment by uga-rosa** [comment]
Implementation is [here](https://github.com/uga-rosa/ccc.nvim/blob/main/lua/ccc/picker/hex.lua).
In fact, `lua/ccc/picker/*` are all reference implementations, and users can create and specify their own pickers as they wish.

**Comment by uga-rosa** [comment]
https://github.com/uga-rosa/ccc.nvim/pull/122

New solution.
